### ISSUE TITLE
See the full map

### ISSUE THREAD

**Issue by Loobinex** [OpenDungeons/OpenDungeons#1151]
I would like to have an overview of the full map. The levels are quite large and I find I'm always looking for the area's I'm interested in, like where I was digging or where there's still gold to mine.

Here I want to be able to see the entire level and be able to click on a specific location to jump to it.

**Comment by hwoarangmy** [comment]
IMHO, the minimap should show the whole map. ATM, that's not how it is done but it would be better to see everything

@OpenDungeons/developers WDYT ?

**Comment by eugeneloza** [comment]
At this moment I don't really see the minimap useful. So, it really might show the whole map.

But if it will look like the current minimap (top-down full render view) it might cause lags and on large maps it'll be "unreadable". AFAIR the minimap was more simple and large-scale in previous versions, but it was remade due to lags it caused.
Maybe, changing (and maybe, even customizing) the scale would be a good (temporary) solution.

However, to make the minimap useful and efficient, I think it should use simplified render and be more schematic (like the previous minimap).

**Comment by Danimal696** [comment]
eugeneloza +1

**Comment by hwoarangmy** [comment]
IIRC, the old minimap used 1 pixel per tile. Since its size was hardcoded in the GUI, it means that if the level was bigger, only part of it was visible. I think the best would be to group the tiles per pixel. For example, if the level is 200x200 and the minimap is 100x100, 1 pixel should represent 4 pixels. Then, depending on what is on one of the group of tiles, the pixel should take a different color depending on priority. For example, enemy creature > allied creature > gold stack > dirt > claimed.

This way, we would have something similar to the old minimap but with vision on the whole level.

**Comment by Loobinex** [comment]
When the minimap becomes too small, it becomes useless. Corridors are single tiles wide, so they will disappear when you do that. The 1 pixel per tiles seems better, with a toggle to expand the minimap it to full screen(you can call that the not-so-mini-map).
To keep performance ok you can stop rendering the game screen.

**Comment by eugeneloza** [comment]
link https://github.com/OpenDungeons/OpenDungeons/issues/917 & https://github.com/OpenDungeons/OpenDungeons/issues/839

**Comment by Bertram25** [comment]
The old minimap used 4 pixels-wide squares for one tile. I was useful, but looked really ugly as it was using raw colors + the slow use of texture writing each frame.

To have a more schematic minimap, we should have a tileset of simplified tiles texture representing each terrain type and use that to draw either the full minimap, or a convenient part of it.
We could even use a pretty dumb tileset at first that we could improve once the code is there.

I'm also for adding the options to show the full map, and hide the minimap when desired.

**Comment by Loobinex** [comment]
@hwoarangmy Does this 'minimap' come with an option to toggle it to full screen?

**Comment by hwoarangmy** [comment]
no. It is a different minimap than the 2 other ones. For now, it is not possible to switch minimap.
This one displays all the map and do not rotate with the camera

**Comment by Loobinex** [comment]
Ok, I'll have a look when the binaries are available. I think though that a 'map' is a more important feature to have than a 'minimap'.

**Comment by hwoarangmy** [comment]
what do you call 'map' ? Would it be displaying the minimap bigger ?

**Comment by Loobinex** [comment]
Generally a map is full screen yes. What is key though, is that they have a different purpose.

A map is something you study. You actively open in to plan something or to look for something. This is a separate piece of gameplay altogether.
A minimap is something you keep your eye on during 'regular' gameplay, for example to see if white dots appear that you need to respond to.

A player would open his map at the start of a level to study the layout and plan his dungeon. He'll open it again when for example he is ready to attack, runs low on gold or is notified that a hero party has spawned. Also having a map will stimulate exploration.
When levels are too big and having a full screen map still does not provide enough detail, you can add a 'magnifying glass': A small windows under your cursor that -is- zoomed in enough to provide all the detail the player needs.

**Comment by hwoarangmy** [comment]
Well, we will see with the minimap if it is enough for you to display it bigger or if something else is needed.

Note that the minimap I've added is based on the current drawn minimap and probably lacks many stuff (for example some kind of alert and such). It is more a first draft ^^

**Comment by Loobinex** [comment]
Sure, we'll have a look first.

Having alerts is indeed a feature that fits a minimap. As they are displayed during normal gameplay performance is important to focus on, and they need to be rotated along with the player screen to be useful.

For maps, performance is hardly an issue because the normal gameplay does not need to be displayed. Maps always face north.

For reference, this is how the original game did it:
![map](https://cloud.githubusercontent.com/assets/13840686/17266606/85539fd6-55fa-11e6-8443-c71420273e23.jpg)
![minimap](https://cloud.githubusercontent.com/assets/13840686/17265711/a10cdb4c-55f1-11e6-93ea-7e90972fab9d.jpg)

**Comment by hwoarangmy** [comment]
I've uploaded new binaries with the new minimap. To enable it, you need to go to the 'settings' menu in the 'game' tab

**Comment by Loobinex** [comment]
Sorry, but I've got to say it has become quite useless now. Everything is really small so you can't see what's going on. There's no clear indication on where you are (e.g. a rectangle to indicate where you've got the main screen focused) and clicking on a spot to go there doesn't really work. When I do the screen turns black (it seems to be outside of the level range).
I can't even see the rival portals.

Furthermore, I can also no longer use the T key to navigate to my temple to find my way back.

![whereami](https://cloud.githubusercontent.com/assets/13840686/17270748/e65e6bd8-566a-11e6-999a-6033ed9863bb.jpg)

**Comment by hwoarangmy** [comment]
granted, the rectangle to know where the player is looking is needed (as I said, it is a first draft that only draws the map like the other minimap does). I've checked what happens when clicking the minimap and it seems like there is a bug when the camera is rotated. I will have a look.

**Comment by hwoarangmy** [comment]
I've looked and it is a typo. I should be more carefull when doing last minute changes before committing ^^

**Comment by hwoarangmy** [comment]
I've fixed the binaries to allow clicking on the minimap to move the camera. Note that with the bug, the T shortcut to move back to the temple did work but because the camera goes pretty far, it takes some time. It seems like you were not patient enough ;-)

**Comment by Loobinex** [comment]
@akien-mga: That commit does not fill this request. Could you please re-open this issue?
The request is for a map screen that provides an overview. What @hwoarangmy implemented is a different minimap that is too zoomed out to be useful.

At the very least the map should allow you to see where the portals are located.

**Comment by Loobinex** [comment]
@hwoarangmy, I've tried the updated minimap in 0.7.1 and noticed the rectangle does not quite show what's on screen. (notice the rectangle covers the border rock, but you can't see it.) That makes it a bit confusing.

![rectangle](https://cloud.githubusercontent.com/assets/13840686/18022398/d079b612-6bee-11e6-805c-ca4204d48ebd.jpg)

**Comment by hwoarangmy** [comment]
To know which tiles are on the border, we use ogre to know where the ray on each corner pixel hits the plane Z=0. When the camera is rotated like it is in this case, it probably hits a little too far.

**Comment by Loobinex** [comment]
It never matches in perspective view. I just took it to the border to make it easy to see. Compare the rest of the picture.

Edit, o wait, that's what you said, nevermind.

**Comment by hwoarangmy** [comment]
I guess one easy fix would be to check where the ray hits a little bit higher than Z=0 (but since the tileset is configurable, it is a kind of dirty fix).
We could also check which tile is hit like it is done when the player clicks. But I'm not sure how it would hurt performances.

**Comment by hwoarangmy** [comment]
Removing assignment since I'm done with the minimap.
